inline-r/src/Language/R/Parser.hs

```diff
@@ -0,0 +1,348 @@
+-- |
+-- Copyright:   (c) 2016, AlphaSheets, Inc
+-- Stability:   Experimental
+-- Portability: Portable
+--
+-- The "Parser" module provides an interface to deconstruct SEXP values and
+-- to compose destructors, where cascades of cases would be necessary
+-- otherwise.
+--
+-- In addition this is the the simplest way to get additional attributes
+-- that exists in the structure.
+--
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE DeriveGeneric #-}
+module Language.R.Parser
+  ( Parser(..)
+  , parseOnly
+    -- * Parser interface.
+    -- $interface
+  , somesexp
+  , sexp
+  , with
+    -- * Type guards
+    -- $guards
+  , hexp
+  , null
+  , s4
+  , s3
+  , guardType
+    -- * Queries
+  , typeOf
+  , getS3Class
+    -- * Attributes
+    -- $attributes
+  , someAttribute
+  , attribute
+  , attributes
+  , lookupAttribute
+    -- * Attribute parsers.
+  , names
+  , dim
+  , dimnames
+  , rownames
+    -- * Predefined parsers
+    -- $parsers
+  , factor
+    -- * Helpers
+  , charList
+  , choice
+  , list
+  ) where
+
+import Data.Singletons
+import qualified Foreign.R as R
+
+import           Control.DeepSeq
+import           Control.Applicative
+import           Control.Monad (guard, ap, liftM)
+import           Control.Exception (evaluate)
+import           Foreign hiding (void, with)
+import           Foreign.C.String
+import           H.Prelude hiding (typeOf, hexp)
+import qualified H.Prelude as H
+
+import           Data.Typeable (Typeable)
+import           Data.Functor (void)
+import           Data.Foldable (asum)
+import           Data.Traversable
+import           Data.Maybe (mapMaybe)
+import qualified Data.Vector.SEXP as SV
+import           System.IO.Unsafe
+import           GHC.Generics (Generic)
+
+import Prelude hiding (null)
+
+-- | @p :: Parser s r a@ inspects an 'SomeSEXP' expression.
+-- @p@ is bound to the region @s@ where 'SomeSEXP' is allocated,
+-- so the extracted value will not leak out of the region scope.
+--
+-- Running @p@ yields a result of type @r@, and in case of
+-- success a value of type @a@ is produced.
+--
+-- This parser is a pure function, so if you need to allocate
+-- any object (for example for comparison or lookup) you should
+-- do it before running parser.
+--
+newtype Parser s r a = Parser {
+  runParser :: SomeSEXP s  -- ^ expression to parse
+            -> (a -> r)    -- ^ continuation in case of success
+            -> (ParserError s -> r) -- ^ continuation in case of failure
+            -> r
+  }
```

Comment:
Any reason not to define
```Haskell
newtype Parser s a = Parser {
  runParser :: forall r. SomeSEXP s  -- ^ expression to parse
            -> (a -> r)    -- ^ continuation in case of success
            -> (ParserError s -> r) -- ^ continuation in case of failure
            -> r
  }
```
?
It doesn't look like any of the parsers care of what `r` is.